lib/internal/modules/esm/loader.js

```diff
@@ -4,138 +4,156 @@
-    // signature itself).
-    this._resolve = defaultResolve;
-    // This hook is called after the module is resolved but before a translator
-    // is chosen to load it; the format returned by this function is the name
-    // of a translator.
-    this._getFormat = defaultGetFormat;
-    // This hook is called just before the source code of an ES module file
-    // is loaded.
-    this._getSource = defaultGetSource;
-    // This hook is called just after the source code of an ES module file
-    // is loaded, but before anything is done with the string.
-    this._transformSource = defaultTransformSource;
-    // The index for assigning unique URLs to anonymous module evaluation
-    this.evalIndex = 0;
-  }
+ * the main module and everything in its dependency graph.
+ */
+class ESMLoader {
+  /**
+   * Phase 0 of 2 in ESM loading. These are called once before any modules are
+   * started.
+   * @property {function[]} - A last-first list of pre-loading hooks.
+   */
+  #globalPreloaders = [];
 
-  async resolve(specifier, parentURL) {
-    const isMain = parentURL === undefined;
-    if (!isMain)
-      validateString(parentURL, 'parentURL');
-
-    const resolveResponse = await this._resolve(
-      specifier, { parentURL, conditions: DEFAULT_CONDITIONS }, defaultResolve);
-    if (typeof resolveResponse !== 'object') {
-      throw new ERR_INVALID_RETURN_VALUE(
-        'object', 'loader resolve', resolveResponse);
-    }
+  /**
+   * Phase 2 of 2 in ESM loading.
+   * @property {function[]} - A last-first list of loader hooks.
+   */
+  #loaders = [
+    defaultLoad,
+  ];
 
-    const { url } = resolveResponse;
-    if (typeof url !== 'string') {
-      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
-        'string', 'loader resolve', 'url', url);
-    }
-    return url;
-  }
+  /**
+   * Phase 1 of 2 in ESM loading.
+   * @property {function[]} - A last-first list of resolver hooks.
+   */
+  #resolvers = [
+    defaultResolve,
+  ];
 
-  async getFormat(url) {
-    const getFormatResponse = await this._getFormat(
-      url, {}, defaultGetFormat);
-    if (typeof getFormatResponse !== 'object') {
-      throw new ERR_INVALID_RETURN_VALUE(
-        'object', 'loader getFormat', getFormatResponse);
-    }
+  /**
+   * Map of already-loaded CJS modules to use
+   */
+  cjsCache = new SafeWeakMap();
 
-    const { format } = getFormatResponse;
-    if (typeof format !== 'string') {
-      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
-        'string', 'loader getFormat', 'format', format);
-    }
+  /**
+   * The index for assigning unique URLs to anonymous module evaluation
+   */
+  evalIndex = 0;
+
+  /**
+   * Registry of loaded modules, akin to `require.cache`
+   */
+  moduleMap = new ModuleMap();
+
+  /**
+   * Methods which translate input code or other information into es modules
+   */
+  translators = translators;
 
-    if (format === 'builtin') {
-      return format;
+  static pluckHooks({
```

Comment:
I feel like there's a better name for this, but I couldn't think of one.

Reply:
This seems fine :)